File: tests/test_log.py
# input file
f = open('../da_proc_1.out').read().splitlines()

# number of threads and messages
n_thread, n_msg = [int(t) for t in f[0].split()]

print("Have %d threads and %d messages per thread" % (n_thread, n_msg))

# map thread -> messages
sent_by = {}

# loop over lines
for line in f[1:]:
  # parsing one line
  thread, n, message, x = line.split()

  # sanity check
  assert thread == "thread" and message == "message", "Input format is invalid"

  # converting to int
  n, x = int(n), int(x)

  # filling sent_by
  if n not in sent_by.keys(): sent_by[n] = []
  sent_by[n] += [x]

# sanity check: number of threads
assert sorted(list(sent_by.keys())) == list(range(n_thread)), "Have missing threads"

# checking that all messages were logged
for key, value in sent_by.items():
  assert value == range(n_msg), "Thread %d has missing messages" % key

print("Correct")

File: tests/test_log.cpp
#include <stdio.h>
#include <cassert>
#include <stdlib.h>

int main()
{
  // number of threads
  int n;

  // number of messages
  int m;

  // reading first line
  assert(scanf("%d %d\n", &n, &m) == 2);

  // sequence numbers for each thread
  int* seqs = new int[n];

  // zeroing sequence numbers
  for(int i = 0; i < n; i++)
    seqs[i] = 0;

  // reading the rest
  for(int i = 0; i < n * m; i++)
  {
    // thread id
    int t;

    // message id
    int msg;

    // reading a message
    if(scanf("thread %d message %d\n", &t, &msg) != 2)
    {
        printf("Cannot read next message, sequence = [");
        for(int j = 0; j < n; j++) printf("%d ", seqs[j]);
        printf("]\n");
        exit(1);
    }

    // sanity check
    assert(t < n);

    // checking that the next message is correct
    assert(seqs[t] == msg);

    // incrementing corresponding sequence number
    seqs[t]++;
  }

  printf("CORRECT\n");
}
